README.md
# TRF
**Tandem Repeats Finder:** https://tandem.bu.edu/trf/trf.html  
**Tandem Repeats Database:** http://tandem.bu.edu/cgi-bin/trdb/trdb.exe


## Table of Contents ##
 - [Purpose](#purpose)   
 - [Reference](#reference)
 - [Authors](#authors)
 - [License](#license)
 - [Pre-compiled Versions](#pre-compiled-versions)
 - [Instructions for Compiling](#instructions-for-compiling) 
 - [Quick Start](#quick-start)
 - [Using Command Line Version of Tandem Repeats Finder](#using-command-line-version-of-tandem-repeats-finder)  
 - [TRF Definitions](#trf-definitions)  
 	- [FASTA Format](#fasta-format)
    - [Table Explanation](#table-explanation)
    - [Alignment Explanation](#alignment-explanation)
 - [How does Tandem Repeats Finder work?](#how-does-tandem-repeats-finder-work)  
 - [What's New](#whats-new)

## Purpose ## 
A tandem repeat in DNA is two or more adjacent, approximate copies of a pattern of nucleotides. Tandem Repeats Finder is a program to locate and display tandem repeats in DNA sequences. In order to use the program, the user submits a sequence in FASTA format. There is no need to specify the pattern, the size of the pattern or any other parameter. The output consists of two files: a repeat table file and an alignment file. The repeat table, viewable in a web browser, contains information about each repeat, including its location, size, number of copies and nucleotide content. Clicking on the location indices for one of the table entries opens a second browser page that shows an alignment of the copies against a consensus pattern. The program is very fast, analyzing sequences on the order of .5Mb in just a few seconds. Submitted sequences may be of arbitrary length. Repeats with pattern size in the range from 1 to 2000 bases are detected. 

This material is based upon work supported by the National Science Foundation under Grant No. CCR-9623532


## Reference ##
Benson G. **Tandem repeats finder: a program to analyze DNA sequences.** ***Nucleic Acids Res.*** 1999; 27(2):573–580. [doi:10.1093/nar/27.2.573](https://academic.oup.com/nar/article/27/2/573/1061099)

## Authors ##
Gary Benson  
Yozen Hernandez  
Yevgeniy Gelfand   
Alfredo Rodriguez  

## License ##
Tandem Repeats Finder
Copyright (C) 1999-2020 Gary Benson

TRF is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of
the License, or (at your option) any later version.

TRF is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with TRF.  If not, see <https://www.gnu.org/licenses/>.


## Pre-compiled Versions ##
To obtain current and/or earlier, pre-compiled versions of TRF:
 - locate and click on the `releases` link just above the green clone button near the top of this page
 - check in each of the listed releases for executable files
 

## Instructions for Compiling ##
To compile TRF, you will need a C compiler (e.g., gcc, clang) with the standard library installed.
We have tested compiling and installing TRF under UNIX-based systems (Linux, macOS) and for Windows under Cygwin/MinGW.

Brief instructions (advanced):

```bash
# Check actual version
tar xzvf trf-4.10.0.tar.gz
cd trf-4.10.0
mkdir build
cd build
../configure
make
# To install to system
sudo make install
# To copy binary elsewhere
cp src/trf DESTINATION
```

If you are cloning the repository, replace the first two lines above with:

```bash
git clone https://github.com/Benson-Genomics-Lab/TRF.git
cd TRF
```

Step by step:

- Open a terminal window and change directory to the TRF directory created by a clone of this repository.
- Create a directory named 'build': `mkdir build` and change directory to that
- Type `../configure` in the terminal.
- Type `make` in the terminal.
- If you wish to install the binary, type `sudo make install` in the terminal. Otherwise, simply copy the binary from the `src` directory under `build`.
 
This will:

- compile the code
- place the executable version in the build/src directory.  

The file will be called `trf` (`trf.exe` on Windows). For backwards compatibility with automated scripts exepcting TRF to follow a certain naming scheme, the installation will also create a symbolic link named `trf<version>.<operating system>.exe`.
For example the file on a linux 64 bit operating system for version 4.10.0 will be called `trf4.10.0.linux64.exe`.

## Quick Start ##

The following is a recommended command line to run TRF.  Parameters are explained further below. This assumes the executable has been renamed `trf`.

```bash
trf yourfile.fa 2 5 7 80 10 50 2000
```

## Using Command Line Version of Tandem Repeats Finder ##

Once the program is installed you can run it with no parameters to obtain information on proper usage syntax.

If you installed the program as trf then by typing trf at the command line you will see the following output: 

```bash
Please use: trf File Match Mismatch Delta PM PI Minscore MaxPeriod [options]

Where: (all weights, penalties, and scores are positive)
  File = sequences input file
  Match  = matching weight
  Mismatch  = mismatching penalty
  Delta = indel penalty
  PM = match probability (whole number)
  PI = indel probability (whole number)
  Minscore = minimum alignment score to report
  MaxPeriod = maximum period size to report
  [options] = one or more of the following:
        -m        masked sequence file
        -f        flanking sequence
        -d        data file
        -h        suppress html output
        -r        no redundancy elimination
        -l <n>    maximum TR length expected (in millions) (eg, -l 3 or -l=3 for 3 million)
```
Note the sequence file should be in FASTA format:

```bash
>Name of sequence
aggaaacctgccatggcctcctggtgagctgtcctcatccactgctcgctgcctctccag
atactctgacccatggatcccctgggtgcagccaagccacaatggccatggcgccgctgt
actcccacccgccccaccctcctgatcctgctatggacatggcctttccacatccctgtg
```

The program accepts a minimum of eight parameters. Options can be specified to generate additional files.

The following is a more detailed description of the **parameters**:

- **File:** The sequence file to be analyzed in FASTA format (see for details). Multiple sequence in the same file are allowed.
- **Match, Mismatch, and Delta:** Weights for match, mismatch and indels. These parameters are for Smith-Waterman style local alignment using wraparound dynamic programming. Lower weights allow alignments with more mismatches and indels. A match weight of 2 has proven effective with mismatch and indel penalties in the range of 3 to 7. Mismatch and indel weights are interpreted as negative numbers. A 3 is more permissive and a 7 less permissive. The recomended values for Match Mismatch and Delta are 2, 7, and 7 respectively.
- **PM and PI:** Probabilistic data is available for PM values of 80 and 75 and PI values of 10 and 20. The best performance can be achieved with values of PM=80 and PI=10. Values of PM=75 and PI=20 give results which are very similar, but often require as much as ten times the processing time when compared with values of PM=80 and PI=10.
- **Minscore:** The alignment of a tandem repeat must meet or exceed this alignment score to be reported. For example, if we set the matching weight to 2 and the minimun score to 50, assuming perfect alignment, we will need to align at least 25 characters to meet the minimum score (for example 5 copies with a period of size 5).
- **Maxperiod:** Period size is the program's best guess at the pattern size of the tandem repeat. The program will find all repeats with period size between 1 and 2000, but the output can be limited to a smaller range.
- **-m:** This is an optional parameter and when present instructs the program to generate a masked sequence file. The masked sequence file is a FASTA format file containing a copy of the sequence with every location that occurred in a tandem repeat changed to the letter 'N'. The word "masked" is added to the sequence description line just after the '>' character.
- **-f:** If this option is present, flanking sequence around each repeat is recorded in the alignment file. This may be useful for PCR primer determination. Flanking sequence consists of the 500 nucleotides on each side of a repeat.
- **-d:** A data file is produced if this option is present. This file is a text file which contains the same information, in the same order, as the summary table file, plus consensus pattern and repeat sequences. This file contains no labeling and is suitable for additional processing, for example with a perl script, outside of the program.
- **-h:** suppress HTML output (this automatically switches -d to ON)
- **-l \<n\>:** Specifies that the longest TR array expected in the input is at most n million bp long. The default is 2 (for 2 million). Setting this option too high may result in an error message if you did not have enough availablememory. We have only tested this option uo to value 29.
- **-u:** Prints the help/usage message above
- **-v:** Prints the version information
- **-ngs:** More compact .dat output on multisequence files, returns 0 on success. You may pipe input in with this option using - for file name. Short 50 flanks are appended to .dat output. .dat output actually goes to stdout instead of file. Sequence headers are displayed in output as @header. Only headers containing repeats are shown. 

Using recommended parameters the command line will look something like:  

```bash
 trf yoursequence.txt 2 7 7 80 10 50 500 -f -d -m
```

Once the program starts running it will print update messages to the screen. The word "Done" will be printed when the program finishes.

For single sequence input files there will be at least two HTML format output files, a repeat table file and an alignment file. If the number of repeats found is greater than 120, multiple linked repeat tables are produced. The links to the other tables appear at the top and the bottom of each table. To view the results start by opening the first repeat table file with your web browser. This file has the extension ".1.html". Alignment files can be accessed from the repeat table files. Alignment files end with the ".txt.html" extension.

For input files containing multiple sequences a summary page is produced that links to the output of individual sequences. This file has the extension "summary.html". You should start by opening this file if your input had multiple sequences in the same file. Also note that the output files of individual sequences will have an identifier of the form ".sn." ( n an integer) embedded in the name indicating the index of the sequence in the input file. The identifier is omitted for single sequence input files. 

For more information on the output please see [Table Explanation](#table-explanation) and [Alignment Explanation](#alignment-explanation). 

## TRF Definitions ##

### FASTA Format: ###

The FASTA format is a plain text format which looks something like this:

```
>myseq
AGTCGTCGCTAGCTAGCTAGCATCGAGTCTTTTCGATCGAGGACTAGACTTCTAGCTAGC
TAGCATAGCATACGAGCATATCGGTCATGAGACTGATTGGGCTTTAGCTAGCTAGCATAG
CATACGAGCATATCGGTAGACTGATTGGGTTTAGGTTACC
```

The first line starts with a greater than sign ">" and contains a name or other identifier for the sequence. This is the sequence header and must be in a single line. The remaining lines contain the sequence data. The sequence can be in upper or lower case letters. Anything other than letters (numbers for example) is ignored. Multiple sequences can be present in the same file as long as each sequence has its own header.  

### Table Explanation: ### 

The summary table includes the following information:

1. Indices of the repeat relative to the start of the sequence.
2. Period size of the repeat.
3. Number of copies aligned with the consensus pattern.
4. Size of consensus pattern (may differ slightly from the period size).
5. Percent of matches between adjacent copies overall.
6. Percent of indels between adjacent copies overall.
7. Alignment score.
8. Percent composition for each of the four nucleotides.
9. Entropy measure based on percent composition.

If the output contains more than 120 repeats, multiple linked tables are produced. The links to the other tables appear at the top and bottom of each table.

Note: If you save multiple linked summary table files, use the default names supplied by your browser to preserve the automatic linking.

### Alignment Explanation: ### 

The alignment is presented as follows:

1. In each pair of lines, the actual sequence is on the top and a consensus sequence for all the copies is on the bottom.
2. Each pair of lines is one period except for very small patterns.
3. The 10 sequence characters before and after a repeat are shown.
4. Symbol * indicates a mismatch.
5. Symbol - indicates an insertion or deletion.
6. Statistics refers to the matches, mismatches and indels overall between adjacent copies in the sequence, not between the sequence and the consensus pattern.
7. Distances between matching characters at corresponding positions are listed as distance, number at that distance, percentage of all matches.
8. ACGTcount is percentage of each nucleotide in the repeat sequence.
9. Consensus sequence is shown by itself.
10. If chosen as an option, 500 characters of flanking sequence on each side of the repeat are shown. 

Note: If you save the alignment file, use the default name supplied by your browser to preserve the automatic cross-referencing with the summary table.

## How does Tandem Repeats Finder work? ##

### Probabilistic Model of Tandem Repeats ###

We model *alignment* of two tandem copies of a pattern of length n by a sequence of n independent Bernoulli trials (coin-tosses). The probability of success, *P(Heads)*, which we also call *P<sub>M</sub>* or *matching probability*, represents the *average* percent identity between the copies. Each head in the Bernoulli sequence is interpreted as a match between aligned nucleotides. Each tail is a mismatch, insertion or deletion. A second probability, *P<sub>I</sub>* or *indel probability*, specifies the average percentage of insertions and deletions between the copies. Figure 1 illustrates the underlying idea for the model.

<img src="images_for_readme/figure1.gif" width="500" />

We are interested in the distribution of Bernoulli sequences and the properties of alignments that they represent when dealing with a specific pair *(P<sub>M</sub>, P<sub>I</sub>)*, for example, *(P<sub>M</sub> = .80, P<sub>I</sub> = .10)*. Note that these *conservation parameters* serves as a type of extremal bound, i.e., as a quantitative description of the *most divergent copies* we hope to detect.


### Program Outline ###

The program has *detection* and *analysis* components. The detection component uses a set of statistically based criteria to find *candidate* tandem repeats. The analysis component attempts to produce an alignment for each candidate and if successful gathers a number of statistics about the alignment (percent identity, percent indels) and the nucleotide sequence (composition, entropy measure).

### Detection Component ###

We assume that adjacent copies of any pattern will contain some matching characters in corresponding positions. Just how many matches and how the distance between those matches should vary depend on the fixed values of *P<sub>M</sub>* and *P<sub>I</sub>*. We use statistical criteria to answer these questions.

The algorithm looks for matching nucleotides separated by a common distance *d*, which is not specified in advance. For reasons of efficiency it looks for *runs of k matches*, which we call *k-tuple matches*. A *k-tuple* is a window of *k* consecutive characters from the nucleotide sequence. Matching *k-tuples* are two windows with identical contents and if aligned in the Bernoulli model would produce a run of *k* heads. Because we limit ourselves to *k-tuple matches*, we will *not* detect all matching characters. For example, if *k=6* and two windows contain TCATGT and TCTTGT we will *not* know that there are 5 matching characters because the window contents are not identical. Put in terms of the Bernoulli model, the aligned windows would be represented by the sequence HHTHHH which is *not* a run of 6 heads.

The basic operation of the detection component is illustrated in the figure below. Let *S* be a nucleotide sequence. We select a small integer *k*, for the tuple or window size (*k=5* for example) and keep a list of all possible *k*-length strings (there are *4<sup>k</sup>* for the DNA alphabet {A,C,G,T}) which we call the probes. By sliding the window across the sequence, we determine the probe at each position *i* in *S*. For each probe *p*, we maintain a history list, *H<sub>p</sub>*, of the positions at which *p* occurs.

When a position *i* is added to *H<sub>p</sub>*, we scan *H<sub>p</sub>* for all earlier occurrences of *p*. Let one earlier occurrence be at *j*. Since *i* and *j* are the indices of matching *k-tuples*, the distance *d=i-j* is a *possible pattern size* for a tandem repeat. For the criteria tests, we need information about other *k-tuple* matches at the same distance *d* where the leading tuple occurs in the sequence between *j* and *i*. A *distance list*, *D<sub>d</sub>*, stores this information. It can be thought of as a sliding window of length *d* which keeps track of the positions of matches and their total.

List *D<sub>d</sub>* is updated every time a match at distance *d* is detected. Position *i* of the match is stored on the list and the total is increased. The right end of the window is set to *i* and matches that occurred before *j=i-d* are dropped from the list and subtracted from the total. Lists for other nearby distances are also updated at this time (see [Random Walk Distribution](#random-walk-distribution)), but only to reset their right ends to *i* and remove matches that have been passed by the advancing windows. Information in the updated distance lists is used for the [sum-of-heads](#sum-of-heads) and [apparent-size](#apparent-size-distribution) criteria tests. If both tests are successful, the program moves on to the analysis component.

<img src="images_for_readme/figure2.gif" width="500"/>


### Statistical Criteria ###

The statistical criteria are based on runs of heads in Bernoulli sequences, corresponding to matches detected with the *k-tuples* and stored in the distance lists. The criteria are based on four distributions which depend upon 1) the pattern length, *d*, 2) the matching probability, *P<sub>M</sub>*, 3) the indel probability, *P<sub>I</sub>*, and 4) the tuple size, *k*. For each distribution, we either calculate it with a formula or estimate it using simulation. Then, we select a cutoff value that serves as our criterion.

### Sum of Heads Distribution ###


NEWS
Please see the history.txt file or the TRF "What's New?" page online located at

https://tandem.bu.edu/trf/trf.whatnew.html



src/tr30dat.h
/*
Tandem Repeats Finder
Copyright (C) 1999-2020 Gary Benson

This file is part of the Tandem Repeats Finder (TRF) program.

TRF is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of
the License, or (at your option) any later version.

TRF is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
the GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public
License along with TRF.  If not, see <https://www.gnu.org/licenses/>.
*/

/* Feb. 14th, 1997 */
/* This is a version which contains the narrow band alignment routines
   narrowbnd.c, prscores.c, pairalign.c */

#ifndef TR30DAT_H
#define TR30DAT_H

/* These declarations moved by Yevgeniy Gelfand on Jan 27, 2010  */
/* To have smaller sequences not send results */
/* to disc to improve performance             */
int counterInSeq = 0;

/* uncomment only one platform target identifier */

//#define WINDOWSGUI
//#define WINDOWSCONSOLE
//#define UNIXCONSOLE
//#define UNIXGUI

/* make sure only one platform is defined */
#if ( defined( WINDOWSGUI ) + defined( WINDOWSCONSOLE ) + defined( UNIXGUI ) + \
      defined( UNIXCONSOLE ) ) > 1
#error Only one Platform can be defined in tr30dat.h
#endif

/* make sure at least one platform is defined */
// #if ( defined( WINDOWSGUI ) + defined( WINDOWSCONSOLE ) + defined( UNIXGUI ) + \
//       defined( UNIXCONSOLE ) ) == 0
// #pragma message( \
//   "You forgot to define a platform when compiling. Setting UNIXCONSOLE." )
#if __unix__
#define UNIXCONSOLE
#undef WINDOWSGUI
#undef WINDOWSCONSOLE
#elif _WIN32
#define WINDOWSCONSOLE
#undef UNIXCONSOLE
#undef UNIXGUI
#endif

#ifdef UNIXCONSOLE
#define _MAX_PATH 260
#elif defined( UNIXGUI )

/* it is possible to do a windows GTK+ version */
#if !defined( _WIN32 )
#define _MAX_PATH 260  /* max. length of full pathname */
#define _MAX_DIR 260   /* max. length of path component */
#define _MAX_FNAME 260 /* max. length of file name component */
#define _MAX_EXT 260   /* max. length of extension component */
#endif

#endif

/* all include  libraries */
#include <stdio.h>
#include <stddef.h> /* has size_t definition */
#include <stdlib.h> /* has calloc definition */
#include <ctype.h>  /* includes toupper(c) */
#include <string.h> /* includes strncat() */
#include <math.h>   /* for ceil function */
#include <stdarg.h> /* for trf_message() function */

/* use semantic versioning, please: https://semver.org/ */
#ifndef PACKAGE_VERSION
#define versionstring "4.10.0"
#else
#define versionstring PACKAGE_VERSION
#endif

#define farcalloc calloc

#ifndef UNIXGUI
/* alredy defined in glib, used in unix graphical version */
#define TRUE 1
#define FALSE 0
#endif

#define WEIGHTCONSENSUS 0
#define REPEATCONSENSUS 0
#define MAXDISTANCECONSTANT 2000 /* should replaced by a variable, later */
#define MINDISTANCE 10
#define MINBANDRADIUS 6
#define RECENTERCRITERION 3
/* 6/24/05 G. Benson */
//#define MAXBANDWIDTH 300
#define MAXBANDWIDTH 150
#define MAXTUPLESIZES 10
/* 6/17/05 G. Benson */
/* 1/26/10 G. Benson */
/* make MAXWRAPLENGTH an integer instead of a constant */
/* #define MAXWRAPLENGTH 500000 */
/* #define MAXWRAPLENGTHCONST 500000 */
/* 11/17/15 G. Benson */
/* make MAXWRAPLENGTHCONST longer to accomodate long centromeric repeats */
/* #define MAXWRAPLENGTHCONST 5000000 */
/* 01/13/16 Y. Hernandez */
/* make MAXWRAPLENGTHCONST longer to accomodate longer repeat in Human chr 18,
 * HG38 */
/* #define MAXWRAPLENGTHCONST 10000000 */
/* 01/22/16 Y. Hernandez */
/* make MAXWRAPLENGTHCONST longer to accomodate longer repeat in Human chr 18,
 * HG38 */
/* 01/26/16 Y. Hernandez */
/* Let MAXWRAPLENGTHCONST be definable on the command line. Easier to update
 * without changing source. */
/* 02/05/16 Y. Hernandez */
/* End use of MAXWRAPLENGTHCONST macro, use a command line option instead. */
//#ifndef MAXWRAPLENGTHCONST
//#define MAXWRAPLENGTHCONST 10000000
//#endif
/* 02/05/16 Y. Hernandez */
/* Since this is no longer a macro, use all lower case to avoid confusion. */
unsigned int maxwraplength = 0;

/* Added by Yevgeniy Gelfand on Jan 27, 2010  */
/* To have smaller sequences not send results */
/* to disc to improve performance             */
// int TempFilesToDisc = 0;

#define MAXPATTERNSIZECONSTANT                    \
    MAXDISTANCECONSTANT /* replaced by a variable \
                         */
#define DASH '-'
#define BLANK ' '

#define WITHCONSENSUS 1
#define WITHOUTCONSENSUS 0

#define MULTIPLES 3 /* We keep MULTIPLES*d distance entries */
#define FILLMULTIPLE 3
#define LOWERENDMULTIPLE                     \
    3 /* how close the lower end of distance \
         indices has to be to the end of one copy */
#define ACCEPTED 1
#define NOTACCEPTED 0

#define SMALLDISTANCE 20 /* these are sizes for which we do a */
/* full array wraparound dynamic */
/* programming and for which we set d_range */
/* by hand */

int Min_Distance_Entries = 20; /* minimum number of places to store a tuple
                                match.  Usually this is the same as the
                                distance, but for small distances, we
                                allow more tuple matches because we want
                                see a significant number of matches */

int Min_Distance_Window = 20; /* minimum size of distance window. */
/* Usually this is the same as the */
/* distance, but for small distances we */
/* allow more space because we want a */
/* significant number of matches */

#define TAGSEP 50 /* index separation for tags table for linking */
/* active distances for distance range addition of */
/* matches */
int    PM;
int    PI;
double Pindel; /* expected probability of a single character indel in
                  the worst case tandem repeat. Pindel should be tied
                  to the indel cost parameter */

int        MAXDISTANCE    = 500;
int        MAXPATTERNSIZE = 500;
/* */ char debugbuffer[500];

/* G. Benson 1/28/2004 */
/* size of EC increased to avoid memory error when consensus length exceeds
   MAXPATTERNSIZECONSTANT after returning from get_consensus(d) */
/* Y. Hernandez 10/15/2018 */
/* If patternsize over 2000 is ever allowed, must change how this
 * variable is initialized. Must be a dynamically allocated array,
 * should use MAXPATTERNSIZE instead (but set elsewhere, after
 * user parameters have been processed).
 */
unsigned char EC[2 * ( MAXPATTERNSIZECONSTANT + 1 )];

int *Index;
int *ACGTcount;

unsigned char *Sequence;
int            Length;

/* int S[MAXWRAPLENGTH+1][MAXPATTERNSIZE];*/
int    Delta;   /* indel penalty */
int    Alpha;   /* match bonus */
int    Beta;    /* mismatch penalty */
int    AFDelta; /* affine gap initiation penalty */
int    AFGamma; /* affine gap extension penalty */
int    pwidth = 75;
int    Reportmin, Heading;
int    Classlength;
int    Test;
double Rows;
double Totalcharacters;
/* int Lookcount;*/
int  Wrapend;
int  Maxrealrow, Maxrow, Maxcol;
int  Maxscore;
int  ConsClasslength;
int *Tag;    /* list of tags for linking active distances */
int  Toptag; /* last tag in list */

struct pairalign {
    int   length;
    int   score;
    char *textprime, *textsecnd;
    int * indexprime, *indexsecnd;
} AlignPair;

struct cons_data {
    char pattern[2 * ( MAXPATTERNSIZECONSTANT + 1 )];
    int  A[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      C[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      G[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      T[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      dash[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      insert[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      letters[2 * ( MAXPATTERNSIZECONSTANT + 1 )],
      total[2 * ( MAXPATTERNSIZECONSTANT + 1 )];
} Consensus;

/* int Up[MAXPATTERNSIZE+1], Diag[MAXPATTERNSIZE+1];*/

/*******************************************/
/*** new program started 11-29-95 **********/

struct bestperiodlistelement {
    int                           indexhigh;
    int                           indexlow;
    int                           best1;
    int                           best2;
    int                           best3;
    int                           best4;
    int                           best5;
    struct bestperiodlistelement *next;
} Bestperiodlist[1];

struct distanceentry {
    int location;
    int size;
};

struct distancelist {
    int k_run_sums_criteria, waiting_time_criteria, lo_d_range, hi_d_range;
    int numentries, nummatches;
    int lowindex, highindex;
    int linked;
    int linkdown, linkup;
    struct distanceentry *entry;
} * Distance;

#define Lookratio .4

/* created 5/23/05 G. Benson */
/* this array replaces the distanceseenlist.  It stores the extent of
   alignments, both pre- or post-consensus, and is used to block an
   alignment in the same region with the same distance */

struct distanceseenarrayelement {
    int index;
    int end;
    int score;
} * Distanceseenarray;

struct distancelistelement {
    int index;
    int distance;
    int changed_from_distance; /* use for test in
                                  search_for_distance_match_in_distanceseenlist
                                  3/10/05 */
    int end;
    int score;
    int best_possible_score; /* number of copies X length X match weight */
    int accepted;
    struct distancelistelement *next;
} Distanceseenlist[1];
/*******************************************/

/* macros */

#define max4( a, b, c, d )                                                 \
    ( ( a >= b )                                                           \
        ? ( ( a >= c ) ? ( ( a >= d ) ? a : d ) : ( ( c >= d ) ? c : d ) ) \
        : ( ( b >= c ) ? ( ( b >= d ) ? b : d ) : ( ( c >= d ) ? c : d ) ) )
/* returns max of 4 in order a,b,c,d */

#define max3( a, b, c ) \
    ( ( a >= b ) ? ( ( a >= c ) ? a : c ) : ( ( b >= c ) ? b : c ) )
/* returns max of 3 in order a,b,c */

//#define match(a, b) ((a==b)?Alpha:Beta)
/* returns match mismatch matrix value */

/* Jan 27, 2006, Gelfand, changed to use Similarity Matrix to avoid N matching
 * itself */
/* This function may be called multiple times (for different match/mismatch
 * scores) */
int *SM = NULL;
#define match( a, b ) ( SM[256 * ( ( a ) ) + ( b )] )

#define fill_align_pair( c1, c2, l, i, j ) \
    AlignPair.textprime[l]  = c1;          \
    AlignPair.textsecnd[l]  = c2;          \
    AlignPair.indexprime[l] = i;           \
    AlignPair.indexsecnd[l] = j

#define max( a, b ) ( ( ( a ) >= ( b ) ) ? ( a ) : ( b ) )
#define min( a, b ) ( ( ( a ) <= ( b ) ) ? ( a ) : ( b ) )

double  Copynumber;
double  WDPcount;
double  OUTPUTcount;
int *   Criteria_count;
int *   Consensus_count;
int *   Outputsize_count;
double *Cell_count;
double  Try_waiting_time_count, Fail_waiting_time_count;
double  Cell_total, Wasted_total;
/**********************************************************************/
/* New to 2A */

#define GLOBAL 0
#define LOCAL 1

typedef struct {
    unsigned int match;
    unsigned int mismatch;
    unsigned int indel;
    unsigned int minscore;
    unsigned int maxperiod;
    unsigned int PM;
    unsigned int PI;
    int          datafile;
    int          maskedfile;
    int          flankingsequence;
    unsigned int flankinglength;
    int          HTMLoff;
    int          redundoff;
    int          ngs;
    int          use_stdin;
    unsigned int maxwraplength;

    char  inputfilename[_MAX_PATH]; /* constant defined in stdlib */
    char  outputprefix[_MAX_PATH];
    char  outputdirectory[_MAX_PATH];
    char  outputfilename[_MAX_PATH];
    int   multisequencefile; /* flags if file has more than one sequence */
    int   sequenceordinal;   /* holds seq. index starting on 1 */
    int   outputcount;       /* repeats found */
    int   guihandle;         /* this variable is only used in the GUI version */
    int   running;
    char *endstatus;
    int   percent;
} TRFPARAMSET;

TRFPARAMSET paramset; /* this global controls the algorithm */

/* G. Benson */
/* 1/26/10 */
/* change MAXWRAPLENGTH to MAXWRAPLENGTHCONST so MAXWRAPLENGTH can be used as an
 * int */
/* int Bandcenter[MAXWRAPLENGTH+1]; */
int *Bandcenter = NULL;

/* version 2A changes this */
/* int S[MAXWRAPLENGTH+1][MAXDISTANCE+1];*/
/* int Up[MAXDISTANCE+1], Diag[MAXDISTANCE+1];*/

/* int S[MAXWRAPLENGTH+1][MAXBANDWIDTH+1];*/
int **S;

int Up[MAXBANDWIDTH + 1], Diag[MAXBANDWIDTH + 1];
/* version 2A adds max3 and max2 */
#define max2( a, b ) ( ( a >= b ) ? a : b )
/* returns max of 2 in order a,b */

#define max3( a, b, c ) \
    ( ( a >= b ) ? ( ( a >= c ) ? a : c ) : ( ( b >= c ) ? b : c ) )
/* returns max of 3 in order a,b,c */

int Maxrealcol;

/* new for 2Anewt */

int four_to_the[] = {
  1, 4, 16, 64, 256, 1024, 4096, 16384, 65536, 262144, 1048576 };

/* #define Number_tuple_sizes  4 */
/* #define Number_tuple_sizes  4 */

int NTS; /* number of different tuple sizes to use;
            preset for all distances */
/* int Tuplesize[NTS+1]={0,2,3,5,7};*/ /* what the different sizes are */
/* int Tuplesize[NTS+1]={0,4,5,6,7};*/ /* what the different sizes are */
/* int Tuplesize[NTS+1]={0,3,4,5,7};*/
int Tuplesize[MAXTUPLESIZES + 1];
int Tuplemaxdistance[MAXTUPLESIZES + 1];

/* int Tuplemaxdistance[MAXTUPLESIZES+1]={0,30,80,200,MAXDISTANCE};*/ /* upper
                                                                         distance
                                                                         for
                                                                         each
                                                                         tuplesize
                                                                       */
/* int Tuplemaxdistance[MAXTUPLESIZES+1]={0,29,83,159,MAXDISTANCE};*/
/* int Tuplemaxdistance[MAXTUPLESIZES+1]={0,29,159,MAXDISTANCE};*/
int Tuplecode[MAXTUPLESIZES + 1]; /* this is where the actual tuple codes
                                   encountered at a sequence location
                                   are stored. */

int *Tuplehash[MAXTUPLESIZES + 1]; /* points to last location of code
                                    in history list */

int Historysize[MAXTUPLESIZES + 1]; /* size of history lists */

int Nextfreehistoryindex[MAXTUPLESIZES +
                         1]; /*next free location in history index*/

struct historyentry {
    int location, previous, code;
} * History[MAXTUPLESIZES + 1];

struct distribution_parameters {
    double exp;
    double var;
};

int  Criteria_print      = 0;
int  Meet_criteria_print = 0;
int *Sortmultiples;

/* define NUMBER_OF_PERIODS 3 */ /* returns <= 3 best periods for a repeat */
/* modified 3/25/05 G. Benson */
#define NUMBER_OF_PERIODS 5 /* determines 5 best periods for a repeat */
#define NUMBER_OF_PERIODS_TO_TEST \
    3 /* only test 3 best periods for multiples test */
#define NUMBER_OF_PERIODS_INTO_SORTMULTIPLES \
    5 /* added 5/25/05 for compatibility with bestperiodslist */

struct MDDtype {
    int   distance;
    char *direction;
}; /* MDD[MAXWRAPLENGTH+1][MAXBANDWIDTH+1];*/

int F[MAXBANDWIDTH + 1], Fdistance[MAXBANDWIDTH + 1];

int ldong;

int *Statistics_Distance;

FILE *Fptxt;
FILE *Fpdat;

int Minsize = 1;
int Minscore;
int MaxPeriod;

int Period;

int print_flanking = 0;

#define CTRL_SUCCESS 0
#define CTRL_BADFNAME -1
#define CTRL_BADFORMAT -2
#define CTRL_NOTHINGPROCESSED -3

/* the following structure is used to pass a sequence to the algorithm */
#define MAXSEQNAMELEN 200

typedef struct {
    /* Changed to unsigned Feb 16, 2016 Yozen */
    unsigned int length;
    int          composition[26];
    int          nucleotides;
    char         name[MAXSEQNAMELEN];
    char *       sequence;

} FASTASEQUENCE;

void trf_message( char *format, ... ) {

    va_list argp;

    if ( format == NULL )
        return;

    va_start( argp, format );

    if ( !paramset.HTMLoff )
        vfprintf( Fptxt, format, argp );

    va_end( argp );
}

#endif
